### PR TITLE
Add unused arguments optimization

### PR DESCRIPTION
**Pull request by N-Maas**
Finally, we get a global optimization :smile:

### CODE REVIEW

File: src/main/java/edu/kit/compiler/optimizations/UnusedArgumentsOptimization.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+                type.getNRess() == 0 ?  new Type[] {} : new Type[] {type.getResType(0)}
+        ));
+    }
+
+    private void updateCallee(ArgMapping mapping) {
+        Graph graph = mapping.getFunc().getGraph();
+        BackEdges.enable(graph);
+        var argProj = Util.getArgProj(mapping.getFunc());
+        if (argProj.isPresent()) {
+            for (var edge: BackEdges.getOuts(argProj.get())) {
```
Comment (g4rgoil) [inline src/main/java/edu/kit/compiler/optimizations/UnusedArgumentsOptimization.java:80]: Using `BackEdge`s here should be fine. Just as a note, you could also turn the Arg `Proj` into a tuple with the new arguments and set a new  Arg `Proj` for the Graph.

File: src/main/java/edu/kit/compiler/optimizations/analysis/UnusedArgumentsAnalysis.java [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+import edu.kit.compiler.optimizations.Util;
+import firm.BackEdges;
+import firm.Entity;
+import firm.Graph;
+import firm.Mode;
+import firm.bindings.binding_irnode;
+import firm.nodes.Call;
+import firm.nodes.Node;
+import firm.nodes.Proj;
+import firm.nodes.Start;
```
Comment (larsk21) [inline src/main/java/edu/kit/compiler/optimizations/analysis/UnusedArgumentsAnalysis.java:14]: Unused import

File: src/main/java/edu/kit/compiler/JavaEasyCompiler.java [hunk 3]
```diff
@@ -381,7 +382,7 @@ public static void main(String[] args) {
                 asmOptimizer = new AssemblyOptimizer(List.of());
                 break;
             case Level1:
-                optimizer = new Optimizer(List.of(), List.of(
+                optimizer = new Optimizer(List.of(new UnusedArgumentsOptimization()), List.of(
```
Comment (larsk21) [inline src/main/java/edu/kit/compiler/JavaEasyCompiler.java:385]: Would be prettier in a separate line

**Review by larsk21 — APPROVED** [review]
Does what it's supposed to do. Pretty cool!

**Pull request merged**